src/state.py

```diff
@@ -4,25 +4,122 @@
 """Jenkins States."""
 import dataclasses
 import logging
+import typing
+from datetime import datetime
 
 logger = logging.getLogger(__name__)
 
 
+class CharmConfigInvalidError(Exception):
```

Comment:
I'm a bit conflicted about this, I've had the previous jenkins module separated out with `exceptions.py` and `types.py` and the feedback i've received was to have it within the module to allow for more cohesion, so that implementation was rejected. What would you think about a magic number of 3, where if we go over 3 of typing or exception within a module that the module becomes crowded, we can refactor it out to a separate module?
I think it's a good thing we can bring up to the team as well! 
Thank you for your feedback! :D

Reply:
Maybe that was my feedback? It sounds like me. I know having files for several types/exceptions is common practice in Python, but I would rather strive for cohesion especially if these are only referred by a single class. e.g. a exception is only visible within one class

Reply:
I thought that would be a standard to follow since I saw that on Flask 🤔. I asked the same [here](https://github.com/canonical/github-runner-operator/pull/68/files#r1235389098).

Reply:
Yup, I think I additionally received the same comment from Ben during the external review :D

Reply:
Exceptions should inherit from a single parent for a given charm so that it is easy to handle all exceptions for the charm

Regarding where to put them, in packages it can be useful because the file that defines all exceptions is useful for the users of that package to know all the exceptions that can be raised. In charms it is probably ok to leave them in a module